### PR TITLE
Fixes broken SoftLogout UX for homeservers that support both Password and SSO

### PR DESCRIPTION
**Pull request by ericdecanini**
## Type of change

- [ ] Feature
- [x] Bugfix
- [ ] Technical
- [ ] Other :

## Content

Fixes an error in SoftLogout where if your LoginMode was SsoAndPassword, no form would show up on the soft logout screen.

This was done by adding a new field LoginType in SessionParamsEntity which remembers the method the user used to log in.

## Motivation and context

Merging this PR closes https://github.com/vector-im/element-android/issues/5311

I also took this as an opportunity to refactor some classes around the session creation in order to make them easier to read and more testable & maintainable.

I did try to add tests for this too, but due to hardware constraints I couldn't test them on my local machine and using the CI didn't give me the info I need to properly write them to pass (failing because of MockkException, NoClassDefFoundException, etc) so I removed them in [this commit](https://github.com/vector-im/element-android/pull/5398/commits/63389418853c1db018cc5e868c2c64113361ecc4) and [added an issue](https://github.com/vector-im/element-android/issues/5437) to get back to this as soon as I'm able to.

Lastly, this PR also fixes a bug where SSO login after a soft logout would crash the app. I would believe the SSO journey to be fixed but currently there's no true way to test this with both the client and server knowing that the user had soft logged out. The closest we can achieve is to throw a global error on the client side as per the test instructions below.

I attached a gif showing this journey which ultimately ends back in the soft logout screen (assumably because the global error gets thrown again or there's no log in actually happen because the user is already logged in according to the server) so we can only base this fix off of assumptions here

## Screenshots / GIFs

| Password | SSO |
|---|---|
| image | image |

**SSO Journey:**

SVID_20220309_121237_1

## Tests

There is no current way to properly test this without mocking the situation with a few code changes.

1. In SoftLogoutController above line 155 add this line:
globalErrorReceiver.handleGlobalError(GlobalError.InvalidToken(softLogout = true))

2. Run the app and login. You'll immediately be taken to the Soft Logout screen as shown above (though the server doesn't know of any logout happening, it only shows in the app)

3. Test this with a password account and an SSO account

## Tested devices

- [x] Physical
- [ ] Emulator
- OS version(s): Android 10

## Checklist

- [ ] Changes has been tested on an Android device or Android emulator with API 21
- [ ] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#accessibility
- [x] Pull request is based on the develop branch
- [x] Pull request includes a new file under ./changelog.d. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#changelog
- [ ] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [x] You've made a self review of your PR
- [ ] If you have modified the screen flow, or added new screens to the application, you have updated the test [UiAllScreensSanityTest.allScreensTest()](https://github.com/vector-im/element-android/blob/main/vector/src/androidTest/java/im/vector/app/ui/UiAllScreensSanityTest.kt#L73)

### CODE REVIEW

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/auth/login/DefaultLoginWizard.kt
Comment (BillCarsonFr) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/auth/login/DefaultLoginWizard.kt:101]: Could we call it LoginType.CUSTOMN

File: vector/src/main/java/im/vector/app/features/login/LoginActivity.kt
Comment (BillCarsonFr) [inline vector/src/main/java/im/vector/app/features/login/LoginActivity.kt:256]: Unknown and sso are doing the same thing?

**Review by BillCarsonFr — CHANGES_REQUESTED** [review]
Thx eric, good work. I added a few comments after a static review.

It should be tested against a real setup before beeing merged, please reach out in internal room to see how to setup a sandbox homeserver with SSO and ways to revocate tokens.

From the user point of view, I think that we should probably remember the social login used if possible instead and just show that social login button instead of sending back to the raw fallback. Do you think it's possible?

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/auth/login/DirectLoginTask.kt
Comment (ericdecanini) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/auth/login/DirectLoginTask.kt:80]: This must've been a temporary thing I added. Good catch! Changing this to unknown as its not based on any of our known standard login types

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/auth/registration/DefaultRegistrationWizard.kt
Comment (ericdecanini) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/auth/registration/DefaultRegistrationWizard.kt:67]: I thought this too. The ideal solution would be we get this value everytime from the backend but we don't. Another solution would be to persist this either in storage or statically but I think that's more confusing

File: vector/src/main/java/im/vector/app/features/login/LoginActivity.kt
Comment (ericdecanini) [reply vector/src/main/java/im/vector/app/features/login/LoginActivity.kt:256]: Unintentional. Good catch!

**Comment by ericdecanini** [comment]
> Thx eric, good work. I added a few comments after a static review.
> 
> It should be tested against a real setup before beeing merged, please reach out in internal room to see how to setup a sandbox homeserver with SSO and ways to revocate tokens.
> 
> From the user point of view, I think that we should probably remember the social login used if possible instead and just show that social login button instead of sending back to the raw fallback. Do you think it's possible?

Thanks for the comments, I pushed some changes addressing them.

Regarding testing against a real setup, I agree that this would be ideal to do before merging. I'll try tomorrow if I can get this sandbox homeserver set up.

As for remembering social login used, I tried this but this needs much more consideration from a technical standpoint. We could persist which social was used in the database but I think this is the wrong way to go about it. I can see in the code that a recover session through SSO has been set up but not fully implemented. I tried to use this but the implementation was too unclear for me to extend

**Comment by Mikhail5555** [comment]
@ericdecanini I tried this out, and it no longer crashes but it also doesnt solve the issue. It now continues in a login loop.
I did find a very easy way of testing it, set `session_lifetime: 1m` an re-login. This will give you a session which expires within one minute.
Screenrecord: https://cloud.mehome.dev/s/2jP7p6wHSzGK6tG

**Comment by ericdecanini** [comment]
> @ericdecanini I tried this out, and it no longer crashes but it also doesnt fully solve the issue. It now continues in a login loop. I did find a very easy way of testing it, set `session_lifetime: 1m` an re-login. This will give you a session which expires within one minute. Screenrecord: https://cloud.mehome.dev/s/2jP7p6wHSzGK6tG
> 
> edit: I just also read your last comment, it seems that the `not actually login in` is somewhat intended. I think this PR already helpes to solve a lot (Aka the not crashing) and the login issue itself might ned fixing in a seperate PR.

Thanks for testing it! Yeah setting a low `session_lifetime` is exactly the way to test this, and your screen record does give very much good insight. Unfortunately I'm still struggling to look into this deeper as setting up a test sso server on our side is taking ages (you'd think it would be quicker for us). Is there any way you could provide a test sso account on your server, or allow me to register one of my own on your server? Or would you rather we continue trying to set up our own server to test this? (I can't guarantee how quick that would happen though)

**Comment by Mikhail5555** [comment]
@ericdecanini can you contact me on matrix at `@mikhail:mehome.dev`, I will do my best to help you where ever i can, and maybe just setup a second instance with SSO.

**Comment by Mikhail5555** [comment]
@ericdecanini I managed to setup a test server for you which i can provide you with a test account. With the only (major) config difference being the the session time is extreme short (1m). Please contact me through matrix and we can exchange some credentials and other info ;)

**Comment by ericdecanini** [comment]
It must be known that nobody in the team has the capacity at the moment to take care of the issue of SSO login after soft logout causing an infinite loop. After a discussion with @bmarty we determined that this is at least a better outcome than the previous crashing the app had (which also covered soft logout with password which should now be working on this branch) so we'll work towards getting this merged.

We can then open a new issue to address the sso infinite loop in isolation

File: vector/src/main/java/im/vector/app/features/login/LoginActivity.kt
Comment (bmarty) [reply vector/src/main/java/im/vector/app/features/login/LoginActivity.kt:271]: yes, good catch @BillCarsonFr !

**Review by bmarty — APPROVED** [review]
LGTM, thanks.
@BillCarsonFr can you also review so that we can merge this PR? Thanks!

**Review by BillCarsonFr — CHANGES_REQUESTED** [review]
Ok to merge after rebase

**Comment by ericdecanini** [comment]
Force pushed after failed rebase (resulting in 1.8k files changed)

**Review by BillCarsonFr — APPROVED** [review]

**Pull request merged**
